### DISCUSSION TITLE
High-level description of S2's compression process?

### DISCUSSION THREAD

**Discussion by marklit** [klauspost/compress#733] [Q&A] [answered]
Do you have any high-level documentation on how S2 compresses data? I can't find any papers published by Google on the topic and your implementation appears to be the only one available to the public.

I built a flame graph of `s2c` compressing a file and it reported little more than 90% of the time being spent in `encodeBetterBlockAsm4MB`.

![s2c](https://user-images.githubusercontent.com/359316/211131577-006f4333-3e5d-4a10-9dff-c81ca90976f7.svg)

`encodeblock_amd64.s` has around 18K lines of Assembler. I believe this file is hand-written rather than generated based on the commits being made to it but there aren't any annotations. There are 55 jump points with `_encodeBetterBlockAsm4MB` in their name but these won't be visible to `perf` when running.

S2 is said to detect and skip over data that isn't compressible and the way this is described makes me believe Snappy never had this sort of capability but when I examined Google's Snappy C code I can see there is heuristic match skipping support in their encoder. I suspect this is further enhanced in S2 to the point that it's worth mentioning as its own feature. It would be great to learn what different techniques S2 improves upon compared to Snappy.

**Comment by klauspost** [comment, marked as answer]
> Do you have any high-level documentation on how S2 compresses data? 

Compression formats are defined by their decompression algorithm. S2 is defined by the Snappy format ([blocks](https://github.com/google/snappy/blob/main/format_description.txt)/[frames](https://github.com/google/snappy/blob/main/format_description.txt)) with [3 minor changes](https://github.com/klauspost/compress/tree/master/s2#format-extensions).

You can an implementation in the [Go decoder](https://github.com/klauspost/compress/blob/master/s2/decode_other.go#L21).

> I believe this file is hand-written rather than generated based on the commits being made to it but there aren't any annotations.

The assembler is [generated](https://github.com/klauspost/compress/blob/master/s2/_generate/gen.go) from pseudo-assembler using [avo](https://github.com/mmcloughlin/avo).

> S2 is said to detect and skip over data that isn't compressible and the way this is described makes me believe Snappy never had this sort of capability but when I examined Google's Snappy C code I can see there is heuristic match skipping support in their encoder.

I am not entirely sure what you reference. When I reference snappy, it is the Go package. I believe it is in there after [convincing the devs](https://github.com/golang/snappy/pull/19) a long time ago. I am pretty sure I also was the one who convinced them to add it to the C version, but I can't find that information anywhere. Please point me to where I state that the Go package doesn't have that, since that is outdated.

EDIT: Oh, they mentioned it in the [commit](https://github.com/google/snappy/commit/d53de18799418e113e44444252a39b12a0e4e0cc).

> I suspect this is further enhanced in S2 to the point that it's worth mentioning as its own feature.

It is important because many compression users are hesitant to use compression on already compressed material, since many older compressors become very slow on incompressible input. Here it is no big deal, since the compressor will quickly skip it, so it is important developer information.

To see the current encoders, you can reference the Go versions: [default](https://github.com/klauspost/compress/blob/master/s2/encode_all.go#L64), [better](https://github.com/klauspost/compress/blob/master/s2/encode_better.go#L48), [best](https://github.com/klauspost/compress/blob/master/s2/encode_best.go#L21). The snappy compatible versions are in the same files.

**Reply by marklit** [reply]
> Please point me to where I state that the Go package doesn't have that, since that is outdated.

This is most likely my misunderstanding of a description I read somewhere back in 2021.

Thank you for the above. I learned a lot.

**Reply by klauspost** [reply]
Great! 

You are also welcome to look at my [sheet](https://docs.google.com/spreadsheets/d/1nuNE2nPfuINCZJRMt6wFWhKpToF95I47XjSsc-1rbPQ/edit?usp=sharing) which I try to keep up-to-date comparing various Go compression methods on different data type.

`snappy` is the Golang snappy package. `s2` is ... well... s2 and `s2s` is s2 with snappy compatible output. `lz4` is https://github.com/pierrec/lz4
